### PR TITLE
test: Add integrations for CDK projects local commands (local invoke, local start-lambda, local start-api)

### PR DESCRIPTION
**Pull request by mildaniel**
#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/intrinsic_resolver/intrinsic_property_resolver.py [hunk 1]
```diff
@@ -270,9 +270,8 @@ def resolve_resources_assets(self, ignore_errors):
                                 asset.source_local_image, ignore_errors, parent_function=asset.source_property
                             )
                         else:
-                            asset.source_path = self.intrinsic_property_resolver(
-                                asset.source_path, ignore_errors, parent_function=asset.source_property
-                            )
+                            image_tag = asset.image_tag
+                            asset.source_local_image = f"{logical_id}:{image_tag}"
```
Comment (torresxb1) [inline samcli/lib/intrinsic_resolver/intrinsic_property_resolver.py:274]: what was the issue here or why did this change?

File: tests/integration/testdata/invoke/cdk/python/aws-lambda-function/aws_lambda_function/aws_lambda_function_stack.py [hunk 2]
```diff
@@ -85,9 +102,30 @@ def __init__(self, scope: cdk.Construct, construct_id: str, **kwargs) -> None:
...
+        )
+
+        aws_lambda_python.PythonFunction(
+            scope=self,
+            id="python-function-construct",
+            runtime=_lambda.Runtime.PYTHON_3_8,
+            index="app.py",
+            entry="./lambda_code",
+            handler="handler"
+        )
```
Comment (torresxb1) [inline tests/integration/testdata/invoke/cdk/python/aws-lambda-function/aws_lambda_function/aws_lambda_function_stack.py:131]: does the PythonFunction have any difference from the Function above with id `helloworld-serverless-functio`?

File: tests/integration/local/invoke/test_integrations_cli_cdk.py [hunk 3]
```diff
@@ -202,3 +181,37 @@ def test_invoke_with_parameters_overrides(self):
...
         self.assertEqual(environ["MyRuntimeVersion"], "v0")
         self.assertEqual(environ["EmptyDefaultParameter"], "")
+
+    @pytest.mark.flaky(reruns=3)
+    def test_invoke_returns_expected_results_python_function_construct(self):
+        command_list = self.get_command_list(
+            "AwsLambdaFunctionStack/python-function-construct", event_path=self.event_path
+        )
+
+        process = Popen(command_list, stdout=PIPE, cwd=self.working_dir)
```
Comment (hawflau) [inline tests/integration/local/invoke/test_integrations_cli_cdk.py:191]: We have a `run_command` util function defined [here](https://github.com/aws/aws-sam-cli/blob/b0dc9dad779ab0bc63c5688399429ee947b0c86f/tests/testing_utils.py#L26-L37). Can we all use the that one rather repeating over and over again?

File: tests/integration/local/start_api/start_api_integ_base.py [hunk 4]
```diff
@@ -25,6 +28,7 @@ class StartApiIntegBaseClass(TestCase):
 
     @classmethod
     def setUpClass(cls):
+        cls.cmd = cls.base_command()
```
Comment (hawflau) [inline tests/integration/local/start_api/start_api_integ_base.py:31]: I feel like this should be a property instead of a class attribute

**Review by hawflau — CHANGES_REQUESTED** [review]

File: tests/integration/testdata/invoke/cdk/python/aws-lambda-function/aws_lambda_function/aws_lambda_function_stack.py (hunk 2 above)
Comment (mildaniel) [reply tests/integration/testdata/invoke/cdk/python/aws-lambda-function/aws_lambda_function/aws_lambda_function_stack.py:131]: There are some small differences in behaviour between the constructs. These language-specific Lambda Function constructs also build the source code and the generated Cloud Assembly is a little different. I wanted to include this here for completeness.

File: samcli/lib/intrinsic_resolver/intrinsic_property_resolver.py (hunk 1 above)
Comment (mildaniel) [reply samcli/lib/intrinsic_resolver/intrinsic_property_resolver.py:274]: I ran into a bug when testing image assets with `local invoke`. sam-cli was trying to initialize the container using the `asset.source_path` as the image name which resulted in the container failing to start.

File: tests/integration/local/start_api/test_start_api_cdk.py [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+    Testing general requirements around the Service that powers `sam local start-api`
+    """
+
+    template_path = "testdata/start_api/cdk/python/aws-lambda-function"
+
+    def setUp(self):
+        self.url = "http://127.0.0.1:{}".format(self.port)
+
+    def test_static_directory(self):
+        pass
```
Comment (torresxb1) [inline tests/integration/local/start_api/test_start_api_cdk.py:22]: is something missing here?

File: tests/integration/local/start_lambda/start_lambda_api_integ_base.py [hunk 6]
```diff
@@ -133,3 +144,59 @@ def tearDownClass(cls):
...
+        os.mkdir(cls.scratch_dir)
+        cls.working_dir = tempfile.mkdtemp(dir=cls.scratch_dir)
+        construct_definition_path = cls.test_data_path.joinpath("cdk", "python", "aws-lambda-function")
+        copy_tree(construct_definition_path, cls.working_dir)
+        os.chdir(cls.working_dir)
+        super().setUpClass()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.scratch_dir and shutil.rmtree(cls.scratch_dir, ignore_errors=True)
```
[the remark this answers is not in the document]
Comment (mildaniel) [reply tests/integration/local/start_lambda/start_lambda_api_integ_base.py:164]: Not a typo, this is the standard way we write the tear-downs. If the directory is none, it won't try to delete it.

**Review by torresxb1 — APPROVED** [review]

**Review by hawflau — APPROVED** [review]

**Pull request merged**